dev/financial#6: Button for create/view template contribution on contact summary

jaapjansma opened the pull request:
Overview
----------------------------------------

This PR adds a button to the Recurring Contributions on the contact summary to create a template contribution of if a template contribution exists it shows view template.

This is a another PR instead of #20453 where a template contribution is created automatically. 

Why this PR instead? Because there is no consensus over whether we should automatically create a template contribution. In some cases this is necessary, in other cases it might trouble current accounting and bookkeeping reports. 


Before
----------------------------------------

On the recurring contributions tab on the contact summary no view or create template contribution button was shown.

![Screenshot_20210621_170355](https://user-images.githubusercontent.com/4126292/122784381-b81f4200-d2b2-11eb-9f1b-06fc143e8f89.png)


After
----------------------------------------

On the recurring contributions tab on the contact summary a user sees `View Template` 

![Screenshot_20210621_153937](https://user-images.githubusercontent.com/4126292/122784462-c705f480-d2b2-11eb-9589-d959e31998ae.png)

If a template does not exists yet, it will create one. 

When editing or view a template contribution the following is hidden:

* Status
* Receive date
* Send receipt checkbox/email/receipt date
* Payment details block

Comments
----------------------------------------

The button `view template` is only visible when the payment processor `supportsEditRecurringContribution`. 

See https://lab.civicrm.org/dev/financial/-/issues/6
And https://lab.civicrm.org/dev/core/-/issues/2624

mattwire commented:
test this please

mattwire commented on CRM/Contribute/BAO/ContributionRecur.php:
```diff
@@ -403,6 +403,80 @@ public static function supportsFinancialTypeChange($id) {
+    // relevant values to ensure the activity reflects that.
+    $relatedContact = CRM_Contribute_BAO_Contribution::getOnbehalfIds($mostRecentContribution['id']);
+
+    $templateContributionParams = array();
```
Please use short array syntax `[]`

mattwire commented:
@jaapjansma I left one comment. Also please can you squash commits

jaapjansma commented:
@mattwire I have changed the `array()` to `[]` and squashed the commits.

mattwire commented:
@jaapjansma I tried to test this but when I click on "View Template" and edit the contribution it just edits the existing contribution and does not create a template.

Also, I think we need to hide/remove some of the fields (such as receive_date) that do not apply to templates.

jaapjansma commented:
@mattwire to enable to edit a recurring template you need a payment processor which supports Edit Recurring Amount. 
I used the manual payment processor (which is not available in civicrm by default, therefor I added this one to the database manually).

I have updated this PR with hiding certain fields when viewing or editing an template contribution.

mattwire commented:
test this please

mattwire commented:
@jaapjansma This is not working when there is NOT already a template contribution. In this case clicking "View Template" assigns `is_template` = 0 to the template and it updates the existing contribution.

jaapjansma commented:
@mattwire I have fixed the issue you discovered with the is test = 1 and the creation of template contributions. I also updated the code to reflect the changes made in https://github.com/civicrm/civicrm-core/commit/4df23f2a270c08bc42fd1ff93fcfe318ecdc0e19

mattwire commented:
Thankyou @jaapjansma I tested this and it is now working as expected. Please can you update https://lab.civicrm.org/dev/financial/-/issues/6. This should be released in 5.41
